crates/map/src/cave.cairo

```diff
@@ -0,0 +1,397 @@
+//! Cave generation methods.
+
+// Internal imports
+
+use origami_map::helpers::bitmap::Bitmap;
+use origami_map::helpers::power::TwoPower;
+use origami_map::helpers::seeder::Seeder;
+use origami_map::types::direction::Direction;
+
+// Constants
+
+const MAX_SIZE: u8 = 252;
+
+/// Cave struct.
+#[derive(Destruct)]
+pub struct Cave {
+    pub width: u8,
+    pub height: u8,
+    pub grid: felt252,
+}
+
+/// Errors module.
+pub mod errors {
+    pub const CAVE_INVALID_EXIT: felt252 = 'Cave: invalid exit';
+    pub const CAVE_INVALID_DIMENSION: felt252 = 'Cave: invalid dimension';
+}
+
+/// Implementation of the `CaveTrait` trait for the `Cave` struct.
+#[generate_trait]
+pub impl CaveImpl of CaveTrait {
+    #[inline(always)]
+    fn new(width: u8, height: u8, count: u8, seed: felt252) -> Cave {
+        // [Check] Valid dimensions
+        let size = width * height;
+        assert(width * height <= MAX_SIZE, errors::CAVE_INVALID_DIMENSION);
+        // [Effect] Deposite objects uniformly
+        let seed = Seeder::reseed(seed, seed);
+        let default: u256 = seed.into() / TwoPower::power(252 - size);
+        let mut cave = Cave { width, height, grid: default.try_into().unwrap() };
+        Self::generate(ref cave, count.into());
+        // [Return] Cave
+        cave
+    }
+
+    #[inline]
+    fn is_edge(ref self: Cave, x: u8, y: u8) -> bool {
+        let width: u8 = self.width;
+        let height: u8 = self.height;
+        x == 0 || y == 0 || x == width - 1 || y == height - 1
+    }
+
+    #[inline]
+    fn add(ref self: Cave, position: u8) {
+        // [Check] Exit is not a corner and on edge
+        self.assert_not_corner(position);
+        self.assert_on_edge(position);
+        // [Effect] Dig until it reaches the cave
+        let (x, y) = (position % self.width, position / self.width);
+        let direction: Direction = if x == 0 {
+            Direction::East
+        } else if x == self.width - 1 {
+            Direction::West
+        } else if y == 0 {
+            Direction::North
+        } else {
+            Direction::South
+        };
+        Self::dig(ref self, position, direction);
+    }
+
+    #[inline]
+    fn next(ref self: Cave, position: u8, direction: Direction) -> u8 {
+        let width = self.width;
+        match direction {
+            Direction::North => position + width,
+            Direction::East => position + 1,
+            Direction::South => position - width,
+            Direction::West => position - 1,
+            _ => position,
+        }
+    }
+
+    #[inline]
+    fn dig(ref self: Cave, position: u8, direction: Direction) {
```

Comment:
Sometimes `Direction` is used, and sometimes `u8`, are they not referring to the same? Or it's for convenience?

Reply:
Part of the design I have to rethink, thank you

Reply:
I left Direction type aside since I leverage packed numbers to manage random walks, which is more convenient than enums and avoid useless inferences imo.

```cairo
#[inline]
    fn compute_shuffled_directions(seed: felt252) -> u32 {
        // [Compute] Random number
        let mut random: u32 = (seed.into() % 24_u256).try_into().unwrap();
        // [Return] Pickup a random permutation
        match random {
            0 => 0x0123,
            [...]
            22 => 0x3201,
            _ => 0x3210,
        }
```
([source](https://github.com/dojoengine/origami/pull/90/files#diff-91c2d847825c3f10a852d7d242f4575632dde55fead34c43ea8655de10b84dddR157))